vault/token_store.go

```diff
@@ -1342,16 +1382,49 @@ func (ts *TokenStore) handleTidy(ctx context.Context, req *logical.Request, data
 			// Look up tainted entries so we can be sure that if this isn't
 			// found, it doesn't exist. Doing the following without locking
 			// since appropriate locks cannot be held with salted token IDs.
+			// Also perform deletion if the parent doesn't exist any more.
 			te, _ := ts.lookupSalted(ctx, child, true)
-			if te == nil {
+			// If the child entry is not nil, but the parent doesn't exist, then turn
+			// that child token into an orphan token. Theres no deletion in this case.
+			if te != nil && exists == nil {
+				lock := locksutil.LockForKey(ts.tokenLocks, te.ID)
+				lock.Lock()
+
+				te.Parent = ""
+				err = ts.store(ctx, te)
+				if err != nil {
+					tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to convert child token into an orphan token: %v", err))
+				}
+				lock.Unlock()
+				continue
+			}
+			// Otherwise, if the entry doesn't exist, or if the parent doesn't exist go
+			// on with the delete on the secondary index
+			if te == nil || exists == nil {
 				index := parentPrefix + parent + child
 				ts.logger.Trace("token: deleting invalid secondary index", "index", index)
 				err = ts.view.Delete(ctx, index)
 				if err != nil {
 					tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to delete secondary index: %v", err))
+					continue
+				}
+				deletedChildrenCount++
+			}
+		}
+		// Add current children deleted count to the total count
+		deletedCountParentList += deletedChildrenCount
+		// If we deleted all the children, then add that to our deleted parent entries count
+		if originalChildrenCount == deletedChildrenCount {
+			// Make sure that we only delete if parent doesn't exist.
+			// N.B.: This is a no-op for the consul storage backend since the delete doesn't support
+			// recursive deletes, and delete on a path is a no-op.
+			if exists == nil {
+				err := ts.view.Delete(ctx, parentPrefix+parent)
```

Comment:
@jefferai @vishalnayak I added the explicit delete in here to ensure that the parent prefix gets deleted if there are no children as per previous discussion in [here](https://github.com/hashicorp/vault/pull/4193#discussion_r177172003).

In the case for Consul, it should automatically remove this (and even if it didn't, this would result in a no-op since the underlying ConsulBackend.Delete would have to call DeleteTree to handle tree deletions and deletes on a non-existing key is simply treated as a success). However, other storage backend implementations might error on this since the delete method is usually done against an entry, and not a path/tree (_thinking that this might be an issue for Cassandra's delete, for instance_). This would result in the tidy operation being treated as a failure, so I wonder if we should a) not do this at all or b) treat this as warning instead of an error.

Reply:
@calvn Apologies -- I think this needs to come back out. You're right that we don't know what storage implementations will do, but I believe the expected behavior is that they clean up their own empty prefixes. I remember making this change with `file` for instance, so if a storage backend isn't doing this then it's for the backend to fix. It was right the first way :-/

Reply:
No worries, I'll revert! Glad this discussion happened; I learn something new every day :)